metrics/config.go

```diff
@@ -0,0 +1,126 @@
+/*
+Copyright 2018 The Knative Authors
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package metrics
+
+import (
+	"errors"
+	"fmt"
+	"regexp"
+	"strings"
+
+	"go.uber.org/zap"
+	corev1 "k8s.io/api/core/v1"
+)
+
+const (
+	backendDestinationKey   = "metrics.backend-destination"
+	stackdriverProjectIDKey = "metrics.stackdriver-project-id"
+)
+
+type MetricsBackend string
+
+const (
+	// The metrics backend is stackdriver
+	Stackdriver MetricsBackend = "stackdriver"
+	// The metrics backend is prometheus
+	Prometheus MetricsBackend = "prometheus"
+)
+
+type metricsConfig struct {
+	// The metrics domain. e.g. "serving.knative.dev" or "build.knative.dev".
+	domain string
+	// The component that emits the metrics. e.g. "activator", "autoscaler".
+	component string
+	// The metrics backend destination.
+	backendDestination MetricsBackend
+	// The stackdriver project ID where the stats data are uploaded to. This is
+	// not the GCP project ID.
+	stackdriverProjectID string
+}
+
+var domainRE = regexp.MustCompile("^(serving|build|eventing)\\.knative\\.dev$")
```

Comment:
This feels like a layering violation.  I haven't traced through how this is used, but could we pass this in from the consuming repo?